Create tarball from Alire workspace

mosteo opened the pull request:
A further step towards automating uploading tarballs, when using the `--tar` switch with `alr publish` we create the tarball and ask the user to upload it before continuing.

This is still skinny in the sense that no local checks are performed before the upload, so errors are detected a bit late in the process. Also, no attempt is yet made to detect if the release has a corresponding github release to which the file could be uploaded.

- [x] Implement test
- [x] Document changes

mosteo committed:
Create tarball from alire workspace

mosteo committed:
Tests for tarball creation

mosteo committed:
Workaround for macOS BSD tar

Fabien-Chouteau commented on src/alire/alire-publish.adb:
```diff
@@ -48,10 +50,52 @@ package body Alire.Publish is
+            Log_Success ("Local repository is clean.");
+         when Ahead =>
+            Git_Error ("Your branch is ahead of remote" & ASCII.LF &
+                         "Please push local commits to the remove branch.",
```
`s/remove/remote` That's my own typo by the way :(

Fabien-Chouteau commented on src/alire/alire-publish.adb:
```diff
@@ -227,6 +271,146 @@ package body Alire.Publish is
+                     Target_Dir
+                       / (Milestone
+                          & (if Is_Repo
+                             then ".tgz"
```
Why is there a difference here? (tgz vs tbz2)

mosteo replied:
bzip2 achieves better ratios, but git doesn't support generating tbz2. We could always use .tgz for consistency.

Fabien-Chouteau replied:
It's not a huge deal, I was just curious.

mosteo replied:
OK then, assuming most people will work from git, they wouldn't notice the difference anyway.

Fabien-Chouteau commented on src/alire/alire-publish.adb:
```diff
@@ -227,6 +271,146 @@ package body Alire.Publish is
+      Trace.Always ("Once you have uploaded the file, enter its URL:");
+
+      loop
+         Ada.Text_IO.Put ("Enter URL> ");
```
This could be implemented as a re-usable function in `Alire.Utils.User_Inputs`.

mosteo committed:
Code review fixes

mosteo committed:
Document `alr publish --tar` changes.

mosteo committed:
Fix tar regular expr and improve related test

It seems GNAT's Normalize_Arguments does its own escaping, so we do not need to
do it ourselves. Also, the test checks that indeed the `alire` folder is
excluded when tar'ing manually.

mosteo commented:
I'd say this one is complete, @Fabien-Chouteau. Do you want it merged before branching for the beta?

Fabien-Chouteau commented:
Yes we are not ready to fork yet.

mosteo commented:
Then if you're ok with the changes in the two last new commits I'll merge it.
